runtime/lib/backend/common/CacheManager.hpp

```diff
@@ -102,70 +118,88 @@ class CacheManager {
      */
     void addObservable(const ObsIdType id, const MeasurementsT &callee = MeasurementsT::None)
     {
-        this->obs_keys_.push_back(id);
-        this->obs_callees_.push_back(callee);
+        obs_keys_.push_back(id);
+        obs_callees_.push_back(callee);
     }
 
     /**
      * @brief Get a reference to observables keys.
      */
-    auto getObservablesKeys() -> const std::vector<ObsIdType> & { return this->obs_keys_; }
+    auto getObservablesKeys() -> const std::vector<ObsIdType> & { return obs_keys_; }
 
     /**
      * @brief Get a reference to observables callees.
      */
-    auto getObservablesCallees() -> const std::vector<MeasurementsT> &
-    {
-        return this->obs_callees_;
-    }
+    auto getObservablesCallees() -> const std::vector<MeasurementsT> & { return obs_callees_; }
 
     /**
      * @brief Get a reference to operations names.
      */
-    auto getOperationsNames() -> const std::vector<std::string> & { return this->ops_names_; }
+    auto getOperationsNames() -> const std::vector<std::string> & { return ops_names_; }
 
     /**
      * @brief Get a a reference to operations parameters.
      */
     auto getOperationsParameters() -> const std::vector<std::vector<double>> &
     {
-        return this->ops_params_;
+        return ops_params_;
     }
 
     /**
      * @brief Get a a reference to operations wires.
      */
-    auto getOperationsWires() -> const std::vector<std::vector<size_t>> &
+    auto getOperationsWires() -> const std::vector<std::vector<size_t>> & { return ops_wires_; }
+
+    /**
+     * @brief Get a reference to operations inverses.
+     */
+    auto getOperationsInverses() -> const std::vector<bool> & { return ops_inverses_; }
+
+    /**
+     * @brief Get a reference to operations matrices.
+     */
+    auto getOperationsMatrices() -> const std::vector<std::vector<ComplexT>> &
     {
-        return this->ops_wires_;
+        return ops_matrixs_;
     }
 
     /**
-     * @brief Get a reference to operations inverses.
+     * @brief Get a reference to operations controlled wires.
      */
-    auto getOperationsInverses() -> const std::vector<bool> & { return this->ops_inverses_; }
+    auto getOperationsControlledWires() -> const std::vector<std::vector<size_t>> &
+    {
+        return ops_controlled_wires_;
+    }
+
+    /**
+     * @brief Get a reference to operations controlled values.
+     */
+    auto getOperationsControlledValues() -> const std::vector<std::vector<bool>> &
+    {
+        return ops_controlled_values_;
+    }
 
     /**
      * @brief Get total number of cached gates.
      */
     [[nodiscard]] auto getNumGates() const -> size_t
     {
-        return this->ops_names_.size() + this->obs_keys_.size();
+        return ops_names_.size() + obs_keys_.size();
```

Comment:
Do we need to count matrices here as well?

Reply:
This is already counted in `ops_names_`. I'll add C++ tests for checking this. Thanks!

Reply:
Indeed, I see now, Thanks!